persistent/Database/Persist/Sql/Types.hs

```diff
@@ -114,3 +114,30 @@ type ConnectionPool = Pool SqlBackend
 -- processing).
 newtype Single a = Single {unSingle :: a}
     deriving (Eq, Ord, Show, Read)
+
+-- | An exception indicating that Persistent refused to run some unsafe
+-- migrations. Contains a list of pairs where the Bool tracks whether the
+-- migration was unsafe (True means unsafe), and the Sql is the sql statement
+-- for the migration.
+--
+-- @since 2.11.1.0
+newtype PersistUnsafeMigrationException
+  = PersistUnsafeMigrationException [(Bool, Sql)]
+  deriving Typeable
+
+-- | This 'Show' instance renders an error message suitable for printing to the
+-- console. This is a little dodgy, but since GHC uses Show instances when
+-- displaying uncaught exceptions, we have little choice.
+instance Show PersistUnsafeMigrationException where
+  show (PersistUnsafeMigrationException mig) =
+    concat
+      [ "\n\nDatabase migration: manual intervention required.\n"
+      , "The unsafe actions are prefixed by '***' below:\n\n"
+      , unlines $ map displayMigration mig
+      ]
+    where
+      displayMigration :: (Bool, Sql) -> String
+      displayMigration (True,  s) = "*** " ++ unpack s ++ ";"
+      displayMigration (False, s) = "    " ++ unpack s ++ ";"
```

Comment:
I'd rather have a derived `Show` instance and a custom implementation of `displayException`.

Reply:
I tried that to start with, but unfortunately GHC uses Show, not displayException, when exiting due to an uncaught exception 🤷‍♂️ I’m not sure how else to handle it. Maybe the exception type could just be `data UnsafeMigrationException = UnsafeMigrationException` and we could instead print the human readable error message before throwing it?

Reply:
That’s not super helpful if people do want to catch the error, though.

Reply:
are you serious, GHC? okay, this is fine then

Reply:
haha yeah that was more or less my reaction too